From: Johannes Thumshirn <jthumshirn@suse.de>
Date: Tue, 10 May 2016 15:09:53 +0200
Subject: [PATCH] NVMe: Only release requested regions
References: bsc#979273
Git-commit: edb50a5403d2e2d2b2b63a8365c4378c9c300ed6
Patch-mainline: v4.7-rc3

The NVMe driver only requests the PCIe device's memory regions but releases
all possible regions (including eventual I/O regions). This leads to a stale
warning entry in dmesg about freeing non existent resources.

Signed-off-by: Johannes Thumshirn <jthumshirn@suse.de>
---
 drivers/nvme/host/pci.c |    9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

--- a/drivers/nvme/host/pci.c
+++ b/drivers/nvme/host/pci.c
@@ -1755,9 +1755,14 @@ static int nvme_pci_enable(struct nvme_d
 
 static void nvme_dev_unmap(struct nvme_dev *dev)
 {
+	struct pci_dev *pdev = to_pci_dev(dev->dev);
+	int bars;
+
 	if (dev->bar)
 		iounmap(dev->bar);
-	pci_release_regions(to_pci_dev(dev->dev));
+
+	bars = pci_select_bars(pdev, IORESOURCE_MEM);
+	pci_release_selected_regions(pdev, bars);
 }
 
 static void nvme_pci_disable(struct nvme_dev *dev)
@@ -1994,7 +1999,7 @@ static int nvme_dev_map(struct nvme_dev
 
 	return 0;
 release:
-	pci_release_regions(pdev);
+	pci_release_selected_regions(pdev, bars);
 	return -ENODEV;
 }
 
